src/importsource.cpp

```diff
@@ -83,4 +96,211 @@ ImportSourcePtr ImportSource::clone() const
     return i;
 }
 
+size_t ImportSource::componentCount() const
+{
+    return mPimpl->mComponents.size();
+}
+
+ComponentPtr ImportSource::component(size_t index) const
+{
+    ImportedEntityPtr import = nullptr;
+    if (index < mPimpl->mComponents.size()) {
+        import = mPimpl->mImports.at(mPimpl->mComponents.at(index)).lock();
+    }
+    return std::dynamic_pointer_cast<Component>(import);
+}
+
+bool ImportSource::addComponent(const ComponentPtr &component)
+{
+    auto import = std::dynamic_pointer_cast<ImportedEntity>(component);
+
+    const auto it = std::find_if(mPimpl->mImports.begin(), mPimpl->mImports.end(), [&import](const std::weak_ptr<ImportedEntity> &p) {
+        return p.lock() == import;
+    });
+
+    if (it != mPimpl->mImports.end()) {
+        return false;
+    }
+
+    size_t index = mPimpl->mImports.size();
+    mPimpl->mImports.push_back(import);
+    mPimpl->mComponents.push_back(index);
+    return true;
+}
+
+bool ImportSource::removeComponent(ComponentPtr &component, bool setEmpty)
+{
+    if (component == nullptr) {
+        return false;
+    }
+
+    auto import = std::dynamic_pointer_cast<ImportedEntity>(component);
+    auto it = std::find_if(mPimpl->mImports.begin(), mPimpl->mImports.end(), [&import](std::weak_ptr<ImportedEntity> &p) {
+        return p.lock() == import;
+    });
+
+    if (it == mPimpl->mImports.end()) {
+        return false;
+    }
+
+    mPimpl->removeItem(it);
+
+    if (setEmpty) {
+        ImportSourcePtr empty = nullptr;
+        component->setImportSource(empty);
```

Comment:
As previously agreed/discussed, this should be replaced with `component->setImportSource(nullptr);`. Same below.

Reply:
It's not possible to do this (as I think we did discuss last time!) The compiler complains about the non-const lvalue reference.

Reply:
Oh, that's because there is no `const` in `setImportSource()`. Ok, got it.